### PR TITLE
Bulk form and case reindexer

### PR DESCRIPTION
**Pull request by snopoke**
This is a PR into https://github.com/dimagi/commcare-hq/pull/12271 which should be merged first.

@emord @dimagi/scale-team @millerdev 

best reviewed by commit :fish: 

Depends on https://github.com/dimagi/fakecouch/pull/3

### CODE REVIEW

File: corehq/apps/hqcase/management/commands/ptop_reindexer_v2.py [hunk 1]
```diff
@@ -64,14 +74,12 @@ def confirm():
             return raw_input("Are you sure you want to delete the current index (if it exists)? y/n\n") == 'y'
 
         reindexer = reindex_fns[index]()
+        reindexer_options = clean_options(options)
+        unconsumed = reindexer.consume_options(reindexer_options)
+        if unconsumed:
+            raise CommandError("The following options don't apply to the reindexer you're calling: {}".format(unconsumed.keys()))
```
Comment (dimagimon) [inline corehq/apps/hqcase/management/commands/ptop_reindexer_v2.py]: E501 line too long (129 > 115 characters)

File: corehq/util/couch_doc_processor.py [hunk 2]
```diff
@@ -367,3 +382,47 @@ def _processing_complete(self):
...
             print(DOCS_SKIPPED_WARNING.format(self.skipped))
+
+
+class BulkDocProcessor(CouchDocumentProcessor):
+    """Process couch docs in batches
+
+    The bulk doc processor will send a batch of documents to the document
+    processor. If the processor does not respond with True then
+    the iteration is halted. Restarting the iteration will start by
+    re-sending the previous chunk to the processor
```
Comment (millerdev) [inline corehq/util/couch_doc_processor.py]: I think this statement about re-sending the previous chunk on restart is not always true. Scenario:

- retrieve a doc, we'll call it *the first doc*, `should_process` returns true, doc added to `changes` list
- retrieve number of docs > chunk size; `should_process` happens to return false for all of them. By now the state of the underlying `ResumableDocsByTypeIterator` has advanced and *the first doc* will not be yielded again *unless* it is `retry()`d.
- retrieve enough docs to make `changes` list length == chunk size, `should_process` returns true for all of them
- `_process_chunk` calls `process_bulk_docs`, which returns false and halts the iteration.
- next time the process is run, the first set of changes sent to the processor is not the same as the previous chunk that had been sent.

Comment (millerdev) [inline corehq/util/couch_doc_processor.py]: heh, started writing this before you closed the PR, didn't see you had closed it until after posting.

Comment (snopoke) [inline corehq/util/couch_doc_processor.py]: reopened here https://github.com/dimagi/commcare-hq/pull/12299 - I'll add a test case for that

Comment (snopoke) [inline corehq/util/couch_doc_processor.py]: There's a further issue if multiple doc_types are being used

Comment (snopoke) [inline corehq/util/couch_doc_processor.py]: Fixed in #12299

**Pull request closed without merging**